src/lib.rs

```diff
@@ -303,3 +191,167 @@ fn generate_analysis(config: &Config) -> Result<(), Box<std::error::Error>> {
 
     Ok(())
 }
+
+#[derive(Debug)]
+struct DocData {
+    krate: Crate,
+    data: HashMap<String, Item>,
+}
+
+impl DocData {
+    fn new(config: &Config) -> Result<DocData, Box<std::error::Error>> {
+        let roots = config.host.def_roots()?;
+
+        // FIXME: this whole code shouldn't look for a precise crate.
+        let root_id = roots.iter().find(|&&(_, ref name)| name == "example");
+        let root_id = match root_id {
+            Some(&(id, _)) => id,
+            _ => return Err(Box::new(CrateErr::new("example"))),
+        };
+
+        let root_def = config.host.get_def(root_id)?;
+
+        let name_len = root_def.qualname.len();
+        let mut krate = Crate {
+            id: root_id,
+            // example:: -> example
+            name: root_def.qualname[..(name_len - 2)].to_string(),
+            docs: root_def.docs.clone(),
+            modules: Vec::new(),
+        };
+
+        let data = DocData::build_data(config, root_id, &mut krate)?;
+
+        Ok(DocData { krate, data })
+    }
+
+    fn build_data(
+        config: &Config,
+        root_id: analysis::Id,
+        krate: &mut Crate,
+    ) -> Result<HashMap<String, Item>, Box<std::error::Error>> {
+        let mut data = HashMap::new();
+
+        let defs = config.host.for_each_child_def(
+            root_id,
+            |_, def| def.clone(),
+        )?;
+
+        for def in defs.iter() {
+            match def.kind {
+                DefKind::Mod => {
+                    data.insert(
+                        def.qualname.clone(),
+                        Item::Module {
+                            name: def.name.clone(),
+                            docs: def.docs.clone(),
+                        },
+                    );
+                    krate.modules.push(def.qualname.clone());
+                }
+                DefKind::Static => (),
+                DefKind::Const => (),
+                DefKind::Enum => (),
+                DefKind::Struct => (),
+                DefKind::Union => (),
+                DefKind::Trait => (),
+                DefKind::Function => (),
+                DefKind::Macro => (),
+                DefKind::Tuple => (),
+                DefKind::Method => (),
+                DefKind::Type => (),
+                DefKind::Local => (),
+                DefKind::Field => (),
+            }
+        }
+
+        Ok(data)
+    }
+
+    fn to_json(&self, config: &Config) -> Result<String, Box<std::error::Error>> {
+        use jsonapi::api::*;
+
+        let root_def = config.host.get_def(self.krate.id)?;
```

Comment:
I'm not sure if there is a reason to keep this but it seems we could replace every instance of `root_def` with `krate`. This would also remove the need to pass the `Config` into this function.